### ISSUE TITLE
Improve R executable finding

### ISSUE THREAD

**Issue by renkun-ken** [ManuelHentschel/VSCode-R-Debugger#6]
We could use the method applied in vscode-R to make it more robust to find R executable:

https://github.com/Ikuyadeu/vscode-R/pull/252
https://github.com/Ikuyadeu/vscode-R/pull/340

**Comment by ManuelHentschel** [comment]
Looks good to me :)

**Comment by renkun-ken** [comment]
I'll make a PR for this soon.

**Comment by jmorim** [comment]
I installed the developer branches and get the error "R path not working: [correct path to R.exe]"

I'm using radian, not sure if that affects it, or if it's supposed to use the r terminal.

This is on Windows 10

**Comment by renkun-ken** [comment]
The R path is supposed to be the path to the R terminal. On Windows, we look for R path in registry, then in `PATH`. If the R is installed with default installer settings (Add R path to registry), then user won't need to specify R path.

**Comment by jmorim** [comment]
It's printing the correct path to R when it throws the error so I'm not sure why it's saying the R path isn't working

**Comment by ManuelHentschel** [comment]
On startup, the debugger should print `Starting R session...` to the debug console. That line can be expanded to show something similar to this:
```
v

**Comment by ManuelHentschel** [comment]
On startup, the debugger prints `Starting R session...` to the debug console, which can be expanded to something like this:
```
Starting R session...
  rDelimiter0: <v\s\c>
  rDelimiter1: </v\s\c>
  rPrompt: <#v\s\c>
  rContinue: <##v\s\c>
  rStartup: <v\s\c\R\STARTUP>
  rLibraryNotFound: <v\s\c\LIBRARY\NOT\FOUND>
  rAppend:  ### <v\s\c\COMMAND>
  cwd: undefined
  rPath: "C:\Program Files\R\R-4.0.0\bin\R.exe"
  rArgs: --ess --quiet --interactive --no-save
```
Could you please post the text that is printed for you?

**Comment by jmorim** [comment]
`Starting R session...
rDelimiter0: <v\s\c>
rDelimiter1: </v\s\c>
rPrompt: <#v\s\c>
rContinue: <##v\s\c>
rStartup: <v\s\c\R\STARTUP>
rLibraryNotFound: <v\s\c\LIBRARY\NOT\FOUND>
rAppend:  ### <v\s\c\COMMAND>
cwd: D:\Users\morimoto\Documents\programming\r\twoDxc
rPath: "D:\Program Files\R\R-4.0.0\bin\R.exe"
rArgs: --ess --quiet --interactive --no-save
WARNING: unknown option '--interactive'
R path not working:
"D:\Program Files\R\R-4.0.0\bin\R.exe"`

So it doesn't like the --interactive argument?

**Comment by ManuelHentschel** [comment]
I get the warning about `--interactive`, too, but that should just be ignored by R (the argument is used on linux).

The extension waits 1s for R to be ready, is your R startup time ever longer than that? Or are there any scripts in your .Rprofile that might cause problems?

**Comment by jmorim** [comment]
There shouldn't be, I haven't edited the .Rprofile at all after installing 4.0

**Comment by ManuelHentschel** [comment]
Did you install the extension from the `.vsix` file or are you running it from code?

**Comment by jmorim** [comment]
From .vsix because installing from code gave a different error.

**Comment by ManuelHentschel** [comment]
I'm a bit out of ideas to be honest... Running the extension from code is a bit more complicated than suggested in the readme. And it's rather difficult to debug the extension from the .vsix file, since the unfiltered output of the childprocess (that's actually running R) is currently not output anywhere.

I am currently moving a lot of code around in the branch "setVar" (both the R package and the VSCode extension). You could try installing those versions, maybe you're lucky and that fixes the problem somehow...

**Comment by renkun-ken** [comment]
What if you manually run `D:\Program Files\R\R-4.0.0\bin\R.exe` with `--ess --quiet --interactive --no-save` args? Does it start an R session normally?

**Comment by jmorim** [comment]
I still get "WARNING: unknown option '--interactive', but it takes a couple of seconds to start up, looks like because of whatever packages I had loaded before closing last. I'll try to see if a clean start fixes it

**Comment by renkun-ken** [comment]
@ManuelHentschel Maybe we should let the R session write to a file after startup, and we use `fs.watch` to watch the file to know that the session is properly started.

**Comment by ManuelHentschel** [comment]
Make a file (if not already present) `./.vscode/launch.json` and copy the following content: 
``` json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "R-Debugger",
            "request": "launch",
            "name": "R Debug File",
            "workingDirectory": "${fileDirname}",
            "file": "${file}",
            "debugMode": "file",
            "allowGlobalDebugging": true
        }
    ]
}
```

After saving, it should work to focus a file and press F5 to run it.

**Comment by ManuelHentschel** [comment]
I guess, a quick fix for this would be to add an option to configure the timeout.

Or maybe we could (on the first run) launch R synchronously and check the exit code, before starting it again (this time asynchronously) for the actual debugging.

**Comment by renkun-ken** [comment]
>I believe some form of timeout is necessary here, since the process is started asynchronously, so we do not get an error code if it fails to open R.

Yes exactly. Maybe we should give more tolerance of time since we already captures process error?

**Comment by renkun-ken** [comment]
>I guess, a quick fix for this would be to add an option to configure the timeout.

This sounds good to me.

**Comment by jmorim** [comment]
> Make a file (if not already present) `./.vscode/launch.json` and copy the following content:
> 
> ```json
> {
>     "version": "0.2.0",
>     "configurations": [
>         {
>             "type": "R-Debugger",
>             "request": "launch",
>             "name": "R Debug File",
>             "workingDirectory": "${fileDirname}",
>             "file": "${file}",
>             "debugMode": "file",
>             "allowGlobalDebugging": true
>         }
>     ]
> }
> ```
> 
> After saving, it should work to focus a file and press F5 to run it.

This seems to work. I noticed it was different from whatever was in the debugger config by default but didn't check what the differences were.

**Comment by ManuelHentschel** [comment]
Thanks for your feedback @jmorim !

I tried fixing the issues in the above mentioned pr. If there are no issues with the pr (@renkun-ken ?), I will merge it to develop later today.

**Comment by Trepon** [comment]
Hello,
I also get the error "R path not working: C:\Program Files\R\R-4.0.2\bin"
There seems to be a problem with the space in the path:
```
Starting R session...
R Startup:
{
  "path": "C:\\Program Files\\R\\R-4.0.2\\bin",
  "args": [
    "--ess",
    "--quiet",
    "--no-save"
  ],
  "useJsonServer": true,
  "useSinkServer": true,
  "logLevelCP": 3
}
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
R path not working:
C:\Program Files\R\R-4.0.2\bin
```
This is on Windows 10 with the dev branch, installed from .vsix.

**Comment by ManuelHentschel** [comment]
On windows, the path needs to be the enclosed by `""` and point directly to the `.exe` file. For me the path in the start up info looks as follows:
```
R Startup:
{
  "path": "\"C:\\Program Files\\R\\R-4.0.2\\bin\\R.exe\"",
  "args": [
    "--ess",
    "--quiet",
    "--no-save"
  ],
  "useJsonServer": true,
  "useSinkServer": true,
  "logLevelCP": 3
}
```

If you installed R with the default settings, the program should be able to find the correct path in the registry if you leave the setting `"rdebugger.rterm.windows"` blank.

**Comment by Trepon** [comment]
Thank you for your help, it is working now.
